src/middleware/r2Middleware.ts

```diff
@@ -0,0 +1,138 @@
+import { Request } from '../routes/request';
+import { renderDirectoryListing } from '../utils/directoryListing';
+import { parseConditionalHeaders } from '../utils/request';
+
+// @ts-expect-error lazy loaded
+let provider: R2Provider = undefined;
+
+export class R2Middleware implements Middleware {
+  async handle(request: Request, ctx: Context): Promise<Response> {
+    if (provider === undefined) {
+      provider = new R2Provider({ ctx });
+    }
+
+    const path = getR2Path(request);
+    const isPathADirectory = isDirectoryPath(path);
+
+    // console.log(`path='${path}' directory=${isPathADirectory}`);
+    //  return new Response(`path='${path}' directory=${isPathADirectory} params=${JSON.stringify(request.params)}`);
+
+    return isPathADirectory
+      ? handleDirectory(request, path)
+      : handleFile(request, path);
+  }
+}
+
+async function handleDirectory(
+  request: Request,
+  r2Path: string
+): Promise<Response> {
+  if (!hasTrailingSlash(request.urlObj.pathname)) {
+    return Response.redirect(request.urlObj + '/', 301);
+  }
+
+  const result = await provider.readDirectory(r2Path);
+  if (result === undefined) {
+    return responses.directoryNotFound(request);
+  }
+
+  if (result.hasIndexHtmlFile) {
+    return handleFile(request, r2Path + '/index.html');
+  }
+
+  let lastModified: Date | undefined = undefined;
+  for (const file of result.files) {
+    if (lastModified === undefined || file.lastModified > lastModified) {
+      lastModified = file.lastModified;
+    }
+  }
+
+  let responseBody: string | undefined = undefined;
+  if (request.method === 'GET') {
+    responseBody = renderDirectoryListing(request.urlObj, result);
+  }
+
+  return new Response(responseBody, {
+    headers: {
+      'last-modified': (lastModified ?? new Date()).toUTCString(),
+      'content-type': 'text/html',
+      'cache-control': CACHE_HEADERS.success,
+    },
+  });
+}
+
+function handleFile(request: Request, r2Path: string): Promise<Response> {
+  switch (request.method) {
+    case 'HEAD':
+      return headFile(request, r2Path);
+    case 'GET':
+      return getFile(request, r2Path);
+  }
+
+  throw new Error('R2Middleware handleFile unsupported method');
+}
+
+async function headFile(request: Request, r2Path: string): Promise<Response> {
+  const result = await provider.headFile(r2Path);
+  if (result === undefined) {
+    return responses.fileNotFound(request);
+  }
+
+  return new Response(undefined, {
+    status: result.httpStatusCode,
+    headers: result.httpHeaders,
+  });
+}
+
+async function getFile(request: Request, r2Path: string): Promise<Response> {
+  const result = await provider.getFile(r2Path, {
+    conditionalHeaders: parseConditionalHeaders(request.headers),
+  });
+
+  if (result === undefined) {
+    return responses.fileNotFound(request);
+  }
+
+  return new Response(result.contents, {
+    status: result.httpStatusCode,
+    headers: result.httpHeaders,
+  });
+}
+
+function getR2Path({
```

Comment:
Shouldnt these be handled by itty router why path regular expressions? Ie:

/some/path/{something}/some/other/*matcher*

Reply:
Can't we use itty router again for that? Like a layered strategy?

So an in-router router? Because that is the part I really wish to become "easy to handle" by just using itty-router

Reply:
Maybe? I'm not really sure what that would look like

I think with this approach we're still better off though. It's still a tiny bit messy but a lot more understandable than what we have now and would probably be simpler than having an in-router router imo

I also debated doing something like
```typescript
router.get('/:rootPath/?:path+', [cached(r2Middleware), originMiddleware]);
```

that way we can just use a switch statement in getR2Path, which might help with cleaning it up. Performance wise it should be about the same. Might cause some annoyances if we want to add other GET paths to the worker later down the road though

Reply:
I'm still a -1 with this approach, we doing too much logic, and I believe using an inner router (even if it's not really routing but just path matching feels better.

So we remove the wrangle of if/elses to simply having callback statements;

Hence this function itself instead of returning a value, has callbacks... At least for me it becomes easier to read than all these conditions 😅 

And also easier to maintain as path matchers/globs are easier to maintain imo

Reply:
Let me know, @flakey5 what you decide to do here :)

Reply:
How about we make an issue and take a look at this later on, just to limit this pr from growing any larger 😅 ?

I think I know an approach that would work for this though